src/rgw/driver/d4n/d4n_policy.cc

```diff
@@ -441,20 +508,54 @@ bool LFUDAPolicy::erase_dirty_object(const DoutPrefixProvider* dpp, const std::s
     return false;
   }
 
-  object_heap.erase(p->second->handle);
+  object_heap.erase(p->second.first->handle);
+  p->second.second = State::DONE;
```

Comment:
Here you set the state to DONE and then delete the entry in the following line and then you notify the wait. When the condition for State set to DONE is checked in invalidate_object, do you think that the method will see the correct value of the state? I think the condition must be changed to see if the map entry exists. Think about this and then make the change.

Reply:
I added a log after the entry gets deleted to ensure the state is still accessible and holds the correct value and it seems it does. I'll leave it as is for now but if there are bugs later on during testing, we can keep this in mind as a potential source.

Reply:
Do you think accessing an entry after it has been deleted will hold good? I do not think this is acceptable - if the entry goes away you will be accessing an invalid memory location and will lead to crashes which will be difficult to debug later.

Reply:
I don't think accessing a deleted entry is correct, but the entry itself isn't being accessed in `invalidate_dirty_object`, just the state. I may be thinking of this incorrectly; but the invalidation method is waiting for the `erase_dirty_object` method to change the entry's state to DONE and once this happens, the invalidation method simply returns false. It doesn't try to access the entry or even the state afterwards, so I didn't see a problem with this case occurring. Either way, I will change the condition to that which you suggested but please let me know if I'm misunderstanding the sequence of operations here.

Reply:
Is the state a part of the map entry? And do you erase the map entry in `erase_dirty_object`?

Reply:
And with the map entry being erased after setting the state to DONE, there is no need to maintain a DONE state. Rather the map entry not being present itself shows that the blocks corresponding to the entry have been written to the backend store.